textuilib2.py

```diff
@@ -89,18 +87,14 @@ def clear():
     os.system(clr)
 
 def choice(menu):
-    print ''
-    print 'Choose an Option:'
+    print('')
+    print('Choose an Option:')
     try:
-        option = input('Option: ')
-        i = 0
-        for o in menu.menuItems:
-            if i == option - 1:
-                return o.function
-            i += 1
+        option = int(input('Option: '))
+        if option-1 < len(menu.menuItems):
+            return menu.menuItems[option-1].function
+        else:
+            raise ValueError("Invalid option!")
```

Comment:
I will be changing this line to 'return' rather than 'raise' because I want it to loop back if an invalid option is given.

Reply:
That is not how looping back works. `return` will simply return `None`. You probably want an infinite loop to surround the entire line of code:
```
saneInput = False
while not saneInput:
   <do code>
   ```

Reply:
It doesn't loop that code back. It returns control back to the program which can loop it itself or handle it as needed. The only way I've seen it used is the return and the program handles the loop. I hope that makes SOME sense. I will give an example in a few minutes.

Reply:
Code:
```python
menu1 = menu("Test Menu", [menuItem("Test 1", "test1"), menuItem("Quit", "quit")])

def main():
    option = menu1.displayMenu()

    if option == "test1":
        print("Test 1")
    elif option == "quit":
        sys.exit()
    else:
        main()
```

Program + User Input:
```
textuilib2 - Test Menu

1. Test 1
2. Quit

Option: 3
```

From here, because the user input an invalid response, it should then return control to the program which then loops back into what it was previously doing.

Reply:
While sane in the sense that this won't output any errors, it's a confusing design. The entire point of a menu is to select from a list of predetermined choices, and the point of a library is for an application using the library to not have to worry about the logic of selecting an option.

By inserting looping code into the library itself, the main program doesn't have to worry about anything but the result of that choice:
```
menu1 = menu("Test Menu", [menuItem("Test 1", "test1"), menuItem("Quit", "quit")])

def main():
    option = menu1.displayMenu()

    if option == "test1":
        print("Test 1")
    elif option == "quit":
        sys.exit()
```

In fact, having the program calling displayMenu() handle its own logic can lead to a bug _right now_. Take this code from MLSSEP.py:

```in_file = fileSelect.displayMenu()

    if fileSelect.menuItems == []:
        print("No save files found! Place a .sav file into this directory.")
        sys.exit()

    # print(in_file)
    # time.sleep(5)

    # assert isfile(in_file), "Save file not found!"

    save_data = open(in_file, "rb").read()
    bio = BytesIO(save_data)```

If the library's response to an invalid option is "just return", in_file will be given the returned value of `None` and `open(in_file,"rb")` will crash. If instead displayMenu() looped on its own, this error and others like it could never happen.

A good design principle that python often uses is "let each specific area of the code work separately". In this case, the code that deals with getting and managing user input should ideally be contained in the appropriate library: textlibui2.py. In my opinion, this produces a simpler design for the code, and will lead to less mysterious errors like the one I used as an example.

Reply:
Yes but, with the way the code is designed in the MLSSEP.py file, there is no possibility of there being that bug unless the user deletes a file or tricks the program into believing a file exists that doesn't. The possibility of that error actually happening in practice is 0. Theoretically it's possible as you've proven. However, unless the user or programmer goes out of their way to make the error happen, there isn't a possibility it could happen.

REALIZATION: I just realized that would be a problem if the user entered a non-existent number. However, that can be easily taken care of in the main program.

Reply:
Sorry! Yup, that's what I was trying to say; if an user enters an invalid option, the program would crash. That's why it would be better to just loop in the library's code: to make the calling program not have to check for invalid answers every single time it uses a menu.